### PR TITLE
Improve non-overridden guard optimization in globalVP

### PR DESCRIPTION
**Pull request by gita-omr**
- handle direct calls, use call site index and bytecode index to
  find the right call
- sometimes, at the time the guard is seen not all constraints
  are available for the call. Keep a side table that maps call node
  to the corresponding guard and try to eliminate the guard at the
  time the call is encountered
- support multiple guards pointing to the same block
- Check isTheVirtualCallNodeForAGuardedInlinedCall inside getVirtualCallTreeForGuard

### CODE REVIEW

**Comment by gita-omr** [comment]
Internal tests passed as well after the changes.

File: compiler/optimizer/VPHandlers.cpp [hunk 1]
```diff
@@ -9665,7 +9717,7 @@ static TR::Node *constrainIfcmpeqne(OMR::ValuePropagation *vp, TR::Node *node, b
          }
       }
 
-   if (virtualGuardConstraint && virtualGuardWillBeEliminated)
```
Comment (jdmpapin) [inline compiler/optimizer/VPHandlers.cpp:9668]: Since this makes `virtualGuardConstraint` unused, would you mind removing its declaration as well?

File: compiler/optimizer/VPHandlers.cpp [hunk 2]
```diff
@@ -9429,39 +9494,26 @@ static TR::Node *constrainIfcmpeqne(OMR::ValuePropagation *vp, TR::Node *node, b
...
+         virtualGuardWillBeEliminated = canFoldNonOverriddenGuard(vp, callNode, node);
+         if (!virtualGuardWillBeEliminated &&
+             vp->lastTimeThrough() &&
+             vp->_isGlobalPropagation)
+            {
+            dumpOptDetails(vp->comp(), "Saving guard %p for call %p\n", node, callNode);
+            TR_Pair<TR::TreeTop, TR::CFGEdge> *pair = new (vp->comp()->trHeapMemory())
+                                                   TR_Pair<TR::TreeTop, TR::CFGEdge>(vp->_curTree, edge);
+            List<TR_Pair<TR::TreeTop, TR::CFGEdge>> &list = (vp->_callNodeToGuardNodes)[callNode->getGlobalIndex()];
+            list.setRegion(vp->comp()->trMemory()->heapMemoryRegion());
```
Comment (jdmpapin) [inline compiler/optimizer/VPHandlers.cpp:9508]: Is the stack memory region freed too early? I think this list is only needed within the same pass of GVP

File: compiler/optimizer/VPHandlers.cpp [hunk 3]
```diff
@@ -5400,10 +5400,98 @@ static void devirtualizeCall(OMR::ValuePropagation *vp, TR::Node *node)
...
+         TR_ASSERT_FATAL(guardNode->isTheVirtualGuardForAGuardedInlinedCall() && guardNode->isNonoverriddenGuard(),
+                      "Should be NonoverriddenGuard");
+      
+         if (canFoldNonOverriddenGuard(vp, node, guardNode))
+            {
+            dumpOptDetails(vp->comp(), "can fold this time\n");
+            vp->removeNode(guardNode, false);
+            TR::TransformUtil::removeTree(vp->comp(), treeTop);
+            vp->setEnableSimplifier();
+            vp->_edgesToBeRemoved->add(edge);
```
[the remark this answers is not in the document]
Comment (gita-omr) [reply compiler/optimizer/VPHandlers.cpp:5489]: No, I wanted to use `removeConditionalBranch` but it did not work in this case. I will put it back.

File: compiler/optimizer/OMRValuePropagation.hpp [hunk 4]
```diff
@@ -765,6 +765,8 @@ class ValuePropagation : public TR::Optimization
    //
    TR_Array<TR::CFGEdge *> *_edgesToBeRemoved;
 
+   TR_Array<List<TR_Pair<TR::TreeTop, TR::CFGEdge>>> _callNodeToGuardNodes;
```
[the remark this answers is not in the document]
Comment (gita-omr) [reply compiler/optimizer/OMRValuePropagation.hpp]: Done

File: compiler/optimizer/VPHandlers.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (gita-omr) [reply compiler/optimizer/VPHandlers.cpp:5466]: Done.

File: compiler/optimizer/VPHandlers.cpp (hunk 1 above)
Comment (gita-omr) [reply compiler/optimizer/VPHandlers.cpp:9668]: Done

File: compiler/optimizer/VPHandlers.cpp (hunk 2 above)
Comment (gita-omr) [reply compiler/optimizer/VPHandlers.cpp]: Changed.

**Comment by gita-omr** [comment]
Running tests after the latest changes.

File: compiler/optimizer/ValuePropagationCommon.cpp [hunk 5]
```diff
@@ -266,9 +266,14 @@ void OMR::ValuePropagation::initialize()
    _blocksToBeRemoved = new (trStackMemory()) TR_Array<TR::CFGNode*>(trMemory(), 8, false, stackAlloc);
    _curDefinedOnAllPaths = NULL;
    if (_isGlobalPropagation)
+      {
       _definedOnAllPaths = new (trStackMemory()) DefinedOnAllPathsMap(std::less<TR::CFGEdge *>(), trMemory()->currentStackRegion());
+      _callNodeToGuardNodes = new (trStackMemory()) CallNodeToGuardNodesMap(std::less<TR::Node *>(), trMemory()->currentStackRegion());
+      }
    else
+      {
       _definedOnAllPaths = NULL;
```
Comment (jdmpapin) [inline compiler/optimizer/ValuePropagationCommon.cpp:275]: Should we also null out `_callNodeToGuardNodes` here?

File: compiler/optimizer/VPHandlers.cpp [hunk 6]
```diff
@@ -5400,10 +5400,75 @@ static void devirtualizeCall(OMR::ValuePropagation *vp, TR::Node *node)
...
+   }
+
+
 TR::Node *constrainCall(OMR::ValuePropagation *vp, TR::Node *node)
    {
    constrainChildren(vp, node);
 
+   if (vp->lastTimeThrough() &&
+       vp->_isGlobalPropagation &&
+       (*(vp->_callNodeToGuardNodes))[node])
```
Comment (jdmpapin) [inline compiler/optimizer/VPHandlers.cpp:5443]: Looking up `node` using `operator[]` will insert a null for each absent node queried here. Thankfully it's only the call nodes, but if you want to avoid the insertion, you can use `find()` and compare the result to `end()`. If you keep the result of `find`, that can also eliminate the need for the redundant lookup just below

File: compiler/optimizer/OMRValuePropagation.hpp [hunk 7]
```diff
@@ -80,6 +80,10 @@ extern const ValuePropagationPtr constraintHandlers[];
 typedef TR::typed_allocator<std::pair<TR::CFGEdge * const, TR_BitVector*>, TR::Region &> DefinedOnAllPathsMapAllocator;
 typedef std::map<TR::CFGEdge *, TR_BitVector *, std::less<TR::CFGEdge *>, DefinedOnAllPathsMapAllocator> DefinedOnAllPathsMap;
 
+typedef TR::typed_allocator<std::pair<TR::Node * const, List<TR_Pair<TR::TreeTop, TR::CFGEdge>> *>, TR::Region &> CallNodeToGuardNodesMapAllocator;
+typedef std::map<TR::Node * const, List<TR_Pair<TR::TreeTop, TR::CFGEdge>> *, std::less<TR::Node *>, CallNodeToGuardNodesMapAllocator> CallNodeToGuardNodesMap;
```
Comment (jdmpapin) [inline compiler/optimizer/OMRValuePropagation.hpp:84]: Just curious why the lists are by pointer instead of by value

File: compiler/optimizer/ValuePropagationCommon.cpp (hunk 5 above)
Comment (gita-omr) [reply compiler/optimizer/ValuePropagationCommon.cpp:275]: It will not be used but let me do it for symmetry

File: compiler/optimizer/OMRValuePropagation.hpp (hunk 7 above)
Comment (gita-omr) [reply compiler/optimizer/OMRValuePropagation.hpp]: I was getting some compilation errors. But let me try again...

Comment (gita-omr) [reply compiler/optimizer/OMRValuePropagation.hpp]: Yeah, it complains about List::List() being protected. It worked with TR_Array because TR_Array simply zeros all elements and does not call element constructor.

File: compiler/optimizer/ValuePropagationCommon.cpp (hunk 5 above)
Comment (gita-omr) [reply compiler/optimizer/ValuePropagationCommon.cpp:275]: Done

File: compiler/optimizer/VPHandlers.cpp (hunk 6 above)
Comment (gita-omr) [reply compiler/optimizer/VPHandlers.cpp]: Done

File: compiler/optimizer/OMRValuePropagation.hpp (hunk 7 above)
Comment (gita-omr) [reply compiler/optimizer/OMRValuePropagation.hpp]: Left as List pointer..

**Comment by gita-omr** [comment]
Addressed last 3 comments.

**Comment by fjeremic** [comment]
Jenkins build all

**Comment by fjeremic** [comment]
@jdmpapin any concerns left for this one?

File: compiler/optimizer/OMRValuePropagation.hpp (hunk 7 above)
Comment (jdmpapin) [reply compiler/optimizer/OMRValuePropagation.hpp]: I suspect you still might be able to use it if you use `insert` instead of `operator[]`, but it's not a huge deal :shrug:

**Review by jdmpapin — APPROVED** [review]

File: compiler/optimizer/OMRValuePropagation.hpp (hunk 7 above)
Comment (gita-omr) [reply compiler/optimizer/OMRValuePropagation.hpp]: Oh, ic! I will try to improve later.
